### ISSUE TITLE
Some tests fail on PowerPC [ FAIL 4 | WARN 12 | SKIP 0 | PASS 48 ]

### ISSUE THREAD

**Issue by barracuda156** [kevhuy/WALS#1]
```
R version 4.3.1 (2023-06-16) -- "Beagle Scouts"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: powerpc-apple-darwin10.8.0 (32-bit)

> library(testthat)
> library(WALS)
> 
> test_check("WALS")

finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
finished iteration 6
finished iteration 7
finished iteration 8
algorithm converged

finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
algorithm converged

finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
finished iteration 6
finished iteration 7
finished iteration 8
finished iteration 9
finished iteration 10
finished iteration 11
finished iteration 12
finished iteration 13
finished iteration 14
finished iteration 15
finished iteration 16
finished iteration 17
finished iteration 18
finished iteration 19
finished iteration 20
finished iteration 21
finished iteration 22
finished iteration 23
finished iteration 24
finished iteration 25
finished iteration 26
finished iteration 27
finished iteration 28
finished iteration 29
finished iteration 30
finished iteration 31
finished iteration 32
finished iteration 33
finished iteration 34
finished iteration 35
finished iteration 36
finished iteration 37
finished iteration 38
finished iteration 39
finished iteration 40
finished iteration 41
finished iteration 42
finished iteration 43
finished iteration 44
finished iteration 45
finished iteration 46
finished iteration 47
finished iteration 48
finished iteration 49
finished iteration 50
finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
algorithm converged

finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
algorithm converged

finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
algorithm converged

finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
algorithm converged

finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
algorithm converged

finished iteration 1
finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
algorithm converged

finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
finished iteration 6
algorithm converged

finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
finished iteration 6
algorithm converged

finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
finished iteration 6
finished iteration 7
finished iteration 8
algorithm converged

finished iteration 1
finished iteration 2
finished iteration 3
finished iteration 4
finished iteration 5
finished iteration 6
finished iteration 7
finished iteration 8
algorithm converged

finished iteration 1
finished iteration 1[ FAIL 4 | WARN 12 | SKIP 0 | PASS 48 ]

══ Failed tests ════════════════════════════════════════════════════════════════
── Failure ('test-wals.R:58:3'): Estimates match Magnus et al. (2010), Journal of Econometrics ──
round(coef(fitWals), 4) not identical to `coefVals`.
Objects equal but not identical
── Failure ('test-wals.R:59:3'): Estimates match Magnus et al. (2010), Journal of Econometrics ──
round(sqrt(diag(vcov(fitWals))), 4) not identical to `seVals`.
Objects equal but not identical
── Failure ('test-wals.R:91:3'): Estimates match De Luca & Magnus (2011), The Stata Journal ──
round(coef(fitWals), 7) not identical to `coefVals`.
Objects equal but not identical
── Failure ('test-wals.R:92:3'): Estimates match De Luca & Magnus (2011), The Stata Journal ──
round(sqrt(diag(vcov(fitWals))), 7) not identical to `seVals`.
Objects equal but not identical

[ FAIL 4 | WARN 12 | SKIP 0 | PASS 48 ]
Error: Test failures
Execution halted
```

**Comment by kevhuy** [comment]
Thanks for checking the package on a PowerPC Mac!

I suspect that the tests fail due to differences in the numerical precision of 32 bit vs. 64 bit OSes. As I do not have a 32 bit machine to check, could you please run the following snippet and report the outputs printed in the console?

```
library(WALS)
fit <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
                    law + tropics + avelf + confucian, data = GrowthMPP,
                  prior = laplace(), prescale = FALSE)
print(coef(fit))
print(sqrt(diag(vcov(fit))))

fitScaled <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
                    law + tropics + avelf + confucian, data = GrowthMPP,
                  prior = laplace(), prescale = TRUE)
print(coef(fitScaled))
print(sqrt(diag(vcov(fitScaled))))
```

Another source may also be BLAS/LAPACK as I am unsure how the (I guess self-built?) R version for PowerPC differs from the official releases.

**Comment by barracuda156** [comment]
@kevhuy Thank you for responding! Here is the output:
```
Last login: Mon Sep 25 00:23:10 on ttys002
10:~ svacchanda$ r

R version 4.3.1 (2023-06-16) -- "Beagle Scouts"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: powerpc-apple-darwin10.8.0 (32-bit)

> library(WALS)
> fit <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
+                     law + tropics + avelf + confucian, data = GrowthMPP,
+                   prior = laplace(), prescale = FALSE)
> print(coef(fit))
  (Intercept)        lgdp60      equipinv      school60        life60 
 0.0593870377 -0.0156303659  0.1555043706  0.0174592634  0.0008556707 
    popgrowth           law       tropics         avelf     confucian 
 0.2650796612  0.0146822969 -0.0055319346 -0.0053315028  0.0443342572 
> print(sqrt(diag(vcov(fit))))
 (Intercept)       lgdp60     equipinv     school60       life60    popgrowth 
0.0220668250 0.0032695255 0.0551289696 0.0096967941 0.0003510048 0.2486803492 
         law      tropics        avelf    confucian 
0.0065360482 0.0037126621 0.0048131980 0.0163146968 
> fitScaled <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
+                     law + tropics + avelf + confucian, data = GrowthMPP,
+                   prior = laplace(), prescale = TRUE)
> print(coef(fitScaled))
  (Intercept)        lgdp60      equipinv      school60        life60 
 0.0617513924 -0.0156500724  0.1582128466  0.0166758384  0.0008514925 
    popgrowth           law       tropics         avelf     confucian 
 0.2713868711  0.0134105238 -0.0059973088 -0.0076756667  0.0464550237 
> print(sqrt(diag(vcov(fitScaled))))
 (Intercept)       lgdp60     equipinv     school60       life60    popgrowth 
0.0217908726 0.0031439126 0.0544210220 0.0096670483 0.0003504565 0.2425284505 
         law      tropics        avelf    confucian 
0.0058036514 0.0034556300 0.0050657002 0.0142765330
```

We use `R` from Macports, so the build is transparent, but `BLAS/LAPACK` (and Fortran) may deviate from what is there on x86. (Fortran certainly does in some respects, but should not be relevant here.)
https://github.com/macports/macports-ports/blob/master/math/R/Portfile

P. S. This is likely inconsequential in this case, but I test this in Rosetta (got native PowerPC hardware at home, but away until the end of month).

**Comment by kevhuy** [comment]
@barracuda156 Interesting, the outputs are identical on my Intel Mac. The tests that fail compare the estimated coefs and SEs (after rounding) to results from the literature using `expect_identical()` from `testthat`. Thus, I suspect that either `round()` or `expect_identical()` work differently on PowerPC.

This is probably irrelevant, but just to make sure, could you please check if `coef(fit)` and `sqrt(diag(vcov(fit)))` return vectors? I.e.

```
is.vector(coef(fit))
is.vector(sqrt(diag(vcov(fit))))
```

should both return `TRUE`.

Here is one of the tests where I replaced `expect_identical()` with `expect_equal()`. Could you please check whether this snippet throws any errors on your machine?

```
library(WALS)
library(testthat)

  coefVals <- c("(Intercept)" = .0594,
                "lgdp60" = -.0156,
                "equipinv" = .1555,
                "school60" = .0175,
                "life60" = .0009,
                "popgrowth" = .2651,
                "law" = .0147,
                "tropics" = -.0055,
                "avelf" = -.0053,
                "confucian" = .0443)
  seVals <- c("(Intercept)" = 0.0221,
              "lgdp60" = .0033,
              "equipinv" = .0551,
              "school60" = .0097,
              "life60" = .0004,
              "popgrowth" = .2487,
              "law" = .0065,
              "tropics" = .0037,
              "avelf" = .0048,
              "confucian" = .0163)

  fit <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
                    law + tropics + avelf + confucian, data = GrowthMPP,
                  prior = laplace(), prescale = FALSE)

  expect_equal(round(coef(fit), 4), coefVals)
  expect_equal(round(sqrt(diag(vcov(fit))), 4), seVals)
```

It would also be helpful if you could report the output of `sessionInfo()` after running the tests.

**Comment by barracuda156** [comment]
@kevhuy Vectors are good:
```
> library(WALS)
> fit <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
+                     law + tropics + avelf + confucian, data = GrowthMPP,
+                   prior = laplace(), prescale = FALSE)
> fitScaled <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
+                     law + tropics + avelf + confucian, data = GrowthMPP,
+                   prior = laplace(), prescale = TRUE)
> is.vector(coef(fit))
[1] TRUE
> is.vector(sqrt(diag(vcov(fit))))
[1] TRUE
```
`expect_equal` is silent (no output), but no error:
```
10:~ svacchanda$ r

R version 4.3.1 (2023-06-16) -- "Beagle Scouts"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: powerpc-apple-darwin10.8.0 (32-bit)

> library(WALS)
> library(testthat)
> coefVals <- c("(Intercept)" = .0594,
+                 "lgdp60" = -.0156,
+                 "equipinv" = .1555,
+                 "school60" = .0175,
+                 "life60" = .0009,
+                 "popgrowth" = .2651,
+                 "law" = .0147,
+                 "tropics" = -.0055,
+                 "avelf" = -.0053,
+                 "confucian" = .0443)
> seVals <- c("(Intercept)" = 0.0221,
+               "lgdp60" = .0033,
+               "equipinv" = .0551,
+               "school60" = .0097,
+               "life60" = .0004,
+               "popgrowth" = .2487,
+               "law" = .0065,
+               "tropics" = .0037,
+               "avelf" = .0048,
+               "confucian" = .0163)
> fit <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
+                     law + tropics + avelf + confucian, data = GrowthMPP,
+                   prior = laplace(), prescale = FALSE)
> expect_equal(round(coef(fit), 4), coefVals)
> expect_equal(round(sqrt(diag(vcov(fit))), 4), seVals)
```

```
> sessionInfo()
R version 4.3.1 (2023-06-16)
Platform: powerpc-apple-darwin10.8.0 (32-bit)
Running under: OS X Snow Leopard 10.6.8

Matrix products: default
BLAS:   /opt/local/Library/Frameworks/R.framework/Versions/4.3/Resources/lib/libRblas.dylib 
LAPACK: /opt/local/Library/Frameworks/R.framework/Versions/4.3/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Asia/Jakarta
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] testthat_3.1.10 WALS_0.2.3     

loaded via a namespace (and not attached):
 [1] compiler_4.3.1   magrittr_2.0.3   R6_2.5.1.9000    rprojroot_2.0.3 
 [5] cli_3.6.1.9000   tools_4.3.1      desc_1.4.2       pkgload_1.3.3   
 [9] Formula_1.2-5    brio_1.1.3       rbibutils_2.2.13 rlang_1.1.1.9000
[13] Rdpack_2.4
```

**Comment by kevhuy** [comment]
@barracuda156 Thanks for your quick reply. Sorry, should have asked this before: Could you report the rounded output please?

```
library(WALS)
fit <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
                    law + tropics + avelf + confucian, data = GrowthMPP,
                  prior = laplace(), prescale = FALSE)
print(round(coef(fit), 4))
print(round(sqrt(diag(vcov(fit))), 4))
```
What I get on my (Windows) machine at work:

```
R version 4.3.1 (2023-06-16 ucrt) -- "Beagle Scouts"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: x86_64-w64-mingw32/x64 (64-bit)

> library(WALS)
> fit <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
+                     law + tropics + avelf + confucian, data = GrowthMPP,
+                   prior = laplace(), prescale = FALSE)
> print(round(coef(fit), 4))
(Intercept)      lgdp60    equipinv    school60      life60   popgrowth
     0.0594     -0.0156      0.1555      0.0175      0.0009      0.2651
        law     tropics       avelf   confucian
     0.0147     -0.0055     -0.0053      0.0443
> print(round(sqrt(diag(vcov(fit))), 4))
(Intercept)      lgdp60    equipinv    school60      life60   popgrowth
     0.0221      0.0033      0.0551      0.0097      0.0004      0.2487
        law     tropics       avelf   confucian
     0.0065      0.0037      0.0048      0.0163
```

If your output does not differ, then the culprit should be `expect_identical()` (I do not see any other source).
As a workaround I could replace `expect_identical()` with `expect_equal()` and a small tolerance of `1e-10`. Could you please check if the following runs silently on your machine?


```
library(WALS)
library(testthat)
tol <- 1e-10

# First test
coefVals <- c("(Intercept)" = .0594,
                "lgdp60" = -.0156,
                "equipinv" = .1555,
                "school60" = .0175,
                "life60" = .0009,
                "popgrowth" = .2651,
                "law" = .0147,
                "tropics" = -.0055,
                "avelf" = -.0053,
                "confucian" = .0443)
seVals <- c("(Intercept)" = 0.0221,
              "lgdp60" = .0033,
              "equipinv" = .0551,
              "school60" = .0097,
              "life60" = .0004,
              "popgrowth" = .2487,
              "law" = .0065,
              "tropics" = .0037,
              "avelf" = .0048,
              "confucian" = .0163)

 fit <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
                    law + tropics + avelf + confucian, data = GrowthMPP,
                  prior = laplace(), prescale = FALSE)

 expect_equal(round(coef(fit), 4), coefVals, tolerance = tol)
 expect_equal(round(sqrt(diag(vcov(fit))), 4), seVals, tolerance = 1e-10)

# Second test
coefVals <- c("(Intercept)" = .0617514,
                "lgdp60" = -.0156501,
                "equipinv" = .1582128,
                "school60" = .0166758,
                "life60" = .0008515,
                "popgrowth" = .2713869,
                "law" = .0134105,
                "tropics" = -.0059973,
                "avelf" = -.0076757,
                "confucian" = .046455)
  seVals <- c("(Intercept)" = .0217909,
              "lgdp60" = .0031439,
              "equipinv" = .054421,
              "school60" = .009667,
              "life60" = .0003505,
              "popgrowth" = .2425285,
              "law" = .0058037,
              "tropics" = .0034556,
              "avelf" = .0050657,
              "confucian" = .0142765)

  fitWals <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
                    law + tropics + avelf + confucian, data = GrowthMPP,
                  prior = laplace(), prescale = TRUE)

  expect_equal(round(coef(fitWals), 7), coefVals, tolerance = tol)
  expect_equal(round(sqrt(diag(vcov(fitWals))), 7), seVals, tolerance = tol)

```

Should the checks raise errors, could you please try finding the smallest possible `tol` that makes them pass (run silently)? You could try e.g. `tol <- 1e-9`,  `tol <- 1e-8` ...

**Comment by barracuda156** [comment]
```
10:~ svacchanda$ r

R version 4.3.1 (2023-06-16) -- "Beagle Scouts"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: powerpc-apple-darwin10.8.0 (32-bit)

> library(WALS)
> fit <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
+                     law + tropics + avelf + confucian, data = GrowthMPP,
+                   prior = laplace(), prescale = FALSE)
> print(round(coef(fit), 4))
(Intercept)      lgdp60    equipinv    school60      life60   popgrowth 
     0.0594     -0.0156      0.1555      0.0175      0.0009      0.2651 
        law     tropics       avelf   confucian 
     0.0147     -0.0055     -0.0053      0.0443 
> print(round(sqrt(diag(vcov(fit))), 4))
(Intercept)      lgdp60    equipinv    school60      life60   popgrowth 
     0.0221      0.0033      0.0551      0.0097      0.0004      0.2487 
        law     tropics       avelf   confucian 
     0.0065      0.0037      0.0048      0.0163 
> library(testthat)
> tol <- 1e-10
> coefVals <- c("(Intercept)" = .0594,
+                 "lgdp60" = -.0156,
+                 "equipinv" = .1555,
+                 "school60" = .0175,
+                 "life60" = .0009,
+                 "popgrowth" = .2651,
+                 "law" = .0147,
+                 "tropics" = -.0055,
+                 "avelf" = -.0053,
+                 "confucian" = .0443)
> seVals <- c("(Intercept)" = 0.0221,
+               "lgdp60" = .0033,
+               "equipinv" = .0551,
+               "school60" = .0097,
+               "life60" = .0004,
+               "popgrowth" = .2487,
+               "law" = .0065,
+               "tropics" = .0037,
+               "avelf" = .0048,
+               "confucian" = .0163)
> fit <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
+                     law + tropics + avelf + confucian, data = GrowthMPP,
+                   prior = laplace(), prescale = FALSE)
> expect_equal(round(coef(fit), 4), coefVals, tolerance = tol)
> expect_equal(round(sqrt(diag(vcov(fit))), 4), seVals, tolerance = 1e-10)
> coefVals <- c("(Intercept)" = .0617514,
+                 "lgdp60" = -.0156501,
+                 "equipinv" = .1582128,
+                 "school60" = .0166758,
+                 "life60" = .0008515,
+                 "popgrowth" = .2713869,
+                 "law" = .0134105,
+                 "tropics" = -.0059973,
+                 "avelf" = -.0076757,
+                 "confucian" = .046455)
> seVals <- c("(Intercept)" = .0217909,
+               "lgdp60" = .0031439,
+               "equipinv" = .054421,
+               "school60" = .009667,
+               "life60" = .0003505,
+               "popgrowth" = .2425285,
+               "law" = .0058037,
+               "tropics" = .0034556,
+               "avelf" = .0050657,
+               "confucian" = .0142765)
> fitWals <- wals(gdpgrowth ~ lgdp60 + equipinv + school60 + life60 + popgrowth |
+                     law + tropics + avelf + confucian, data = GrowthMPP,
+                   prior = laplace(), prescale = TRUE)
> expect_equal(round(coef(fitWals), 7), coefVals, tolerance = tol)
> expect_equal(round(sqrt(diag(vcov(fitWals))), 7), seVals, tolerance = tol)
```

Apparently it works fine.

**Comment by kevhuy** [comment]
The rounded results match perfecly, so I do not understand why `expect_identical()` would raise an error. Maybe this is a bug in `expect_identical()` from `testthat`? @barracuda156 Have you encountered similar errors with other R-packages? The package passes all official CRAN checks so it must be something specific about PowerPC or 32bit.

**Comment by barracuda156** [comment]
@kevhuy Yes, I believe we did see this before. In fact, it may be related to macOS SDKs (like, `libm` for example), not the arch, though arch may have an impact too. I recall at least once I saw identical rounding results on `ppc` and `x86_64` which resulted in some test failing.

**Comment by kevhuy** [comment]
Fixed in commit 5c8c57d.

**Comment by barracuda156** [comment]
@kevhuy Thank you!
